# Solar Eclipses

What is the average duration of total darkness during a total solar eclipse? When did the longest solar eclipse occur?

The file `solar-eclipses.csv` provides information on all solar eclipses from 1901 to 2100, leveraging the ability of scientists to accurately predict the timing and location of future eclipses.

A solar eclipse happens when the Moon moves between the Earth and the Sun, blocking the Sun's light either partially or completely. There are three primary types of solar eclipses:

1. Total Solar Eclipse: The Moon fully covers the Sun, casting a shadow on Earth and revealing the Sun's corona.

2. Partial Solar Eclipse: The Moon obscures only part of the Sun, creating a crescent-shaped appearance.

3. Annular Solar Eclipse: The Moon covers the center of the Sun, leaving a ring-like appearance, known as a "ring of fire," around the edges. This occurs when the Moon is too far from Earth to completely cover the Sun.

The `duration` variable indicates the length of time that the entire moon covers the sun.


In [1]:
# FOR GOOGLE COLAB ONLY.
# Uncomment and run the code below. A dialog will appear to upload files.
# Upload 'solar-eclipses.csv'.

# from google.colab import files
# uploaded = files.upload()

In [2]:
import pandas as pd
pd.set_option('display.max_colwidth', 1000)
df = pd.read_csv('solar-eclipses.csv')
df

,date,eclipse_type,magnitude,duration,region
0,05-18-1901,Total,1.068,06m29s,"s Asia, Australia, e Africa [Total: Indonesia, N Guinea, Madagascar]"
1,11-11-1901,Annular,0.922,11m01s,"ne Africa, Asia, w Europe [Annular: ne Africa, India, Sri Lanka, se Asia]"
2,04-08-1902,Partial,0.064,NaN,northern Canada
3,05-07-1902,Partial,0.859,NaN,"New Zealand, South Pacific"
4,10-31-1902,Partial,0.696,NaN,"c Asia, e Europe"
...,...,...,...,...,...
439,10-24-2098,Partial,0.006,NaN,Southern Ocean (near Antarctica)
440,03-21-2099,Annular,0.932,07m32s,"Australia, N.Z., Antarctica, N. America [Annular: Central Pacific]"
441,09-14-2099,Total,1.068,05m18s,"Americas, Africa [Total: Canada, U.S., Central Atlantic]"
442,03-10-2100,Annular,0.934,07m29s,"Australia, N. America [Annular: Central Pacific, U.S.]"


### Project Ideas:

- When did the longest solar eclipse occur? The longest total eclipse?
	- Hint: convert duration to seconds. You can use `str.replace('s', '')` to remove the 's' from the duration values.

- What is the average duration of total solar eclipses? 

- Show the next 10 solar eclipses?
	- Hint: convert date to datetime.


**QUESTION 1. When did the longest solar eclipse occur? The longest total eclipse?
	- Hint: convert duration to seconds. You can use `str.replace('s', '')` to remove the 's' from the duration values.**

In [ ]:
df['duration'].info

<bound method Series.info of 0      06m29s
1      11m01s
2         NaN
3         NaN
4         NaN
        ...  
439       NaN
440    07m32s
441    05m18s
442    07m29s
443    03m32s
Name: duration, Length: 444, dtype: str>

The datatype is str (string). This means we have to convert the data to number.

In [15]:
#making a copy of df (split_duration) with new columns (minutes and seconds)
split_duration = df.copy()

split_duration[['minutes', 'seconds']] = split_duration['duration'].str.split('m', expand=True) 
split_duration

,date,eclipse_type,magnitude,duration,region,minutes,seconds
0,05-18-1901,Total,1.068,06m29s,"s Asia, Australia, e Africa [Total: Indonesia, N Guinea, Madagascar]",06,29s
1,11-11-1901,Annular,0.922,11m01s,"ne Africa, Asia, w Europe [Annular: ne Africa, India, Sri Lanka, se Asia]",11,01s
2,04-08-1902,Partial,0.064,NaN,northern Canada,NaN,NaN
3,05-07-1902,Partial,0.859,NaN,"New Zealand, South Pacific",NaN,NaN
4,10-31-1902,Partial,0.696,NaN,"c Asia, e Europe",NaN,NaN
...,...,...,...,...,...,...,...
439,10-24-2098,Partial,0.006,NaN,Southern Ocean (near Antarctica),NaN,NaN
440,03-21-2099,Annular,0.932,07m32s,"Australia, N.Z., Antarctica, N. America [Annular: Central Pacific]",07,32s
441,09-14-2099,Total,1.068,05m18s,"Americas, Africa [Total: Canada, U.S., Central Atlantic]",05,18s
442,03-10-2100,Annular,0.934,07m29s,"Australia, N. America [Annular: Central Pacific, U.S.]",07,29s


In [19]:
#replacing 's' with '', NaN with 0
split_duration['seconds'] = split_duration['seconds'].str.replace('s', '')
split_duration[['minutes', 'seconds']] = split_duration[['minutes', 'seconds']].fillna(0)
split_duration

,date,eclipse_type,magnitude,duration,region,minutes,seconds
0,05-18-1901,Total,1.068,06m29s,"s Asia, Australia, e Africa [Total: Indonesia, N Guinea, Madagascar]",06,29
1,11-11-1901,Annular,0.922,11m01s,"ne Africa, Asia, w Europe [Annular: ne Africa, India, Sri Lanka, se Asia]",11,01
2,04-08-1902,Partial,0.064,NaN,northern Canada,0,0
3,05-07-1902,Partial,0.859,NaN,"New Zealand, South Pacific",0,0
4,10-31-1902,Partial,0.696,NaN,"c Asia, e Europe",0,0
...,...,...,...,...,...,...,...
439,10-24-2098,Partial,0.006,NaN,Southern Ocean (near Antarctica),0,0
440,03-21-2099,Annular,0.932,07m32s,"Australia, N.Z., Antarctica, N. America [Annular: Central Pacific]",07,32
441,09-14-2099,Total,1.068,05m18s,"Americas, Africa [Total: Canada, U.S., Central Atlantic]",05,18
442,03-10-2100,Annular,0.934,07m29s,"Australia, N. America [Annular: Central Pacific, U.S.]",07,29


In [27]:
#converting minutes and seconds to numeric values & caculating total duration in seconds
split_duration[['minutes', 'seconds']] = split_duration[['minutes', 'seconds']].astype('int')
split_duration['total_duration'] = split_duration['minutes'] * 60 + split_duration['seconds']
split_duration


,date,eclipse_type,magnitude,duration,region,minutes,seconds,total_duration
0,05-18-1901,Total,1.068,06m29s,"s Asia, Australia, e Africa [Total: Indonesia, N Guinea, Madagascar]",6,29,389
1,11-11-1901,Annular,0.922,11m01s,"ne Africa, Asia, w Europe [Annular: ne Africa, India, Sri Lanka, se Asia]",11,1,661
2,04-08-1902,Partial,0.064,NaN,northern Canada,0,0,0
3,05-07-1902,Partial,0.859,NaN,"New Zealand, South Pacific",0,0,0
4,10-31-1902,Partial,0.696,NaN,"c Asia, e Europe",0,0,0
...,...,...,...,...,...,...,...,...
439,10-24-2098,Partial,0.006,NaN,Southern Ocean (near Antarctica),0,0,0
440,03-21-2099,Annular,0.932,07m32s,"Australia, N.Z., Antarctica, N. America [Annular: Central Pacific]",7,32,452
441,09-14-2099,Total,1.068,05m18s,"Americas, Africa [Total: Canada, U.S., Central Atlantic]",5,18,318
442,03-10-2100,Annular,0.934,07m29s,"Australia, N. America [Annular: Central Pacific, U.S.]",7,29,449


In [33]:
#When did the longest solar eclipse occur?
split_duration.sort_values(by='total_duration', ascending=False).head(1)

,date,eclipse_type,magnitude,duration,region,minutes,seconds,total_duration
126,12-14-1955,Annular,0.918,12m09s,"c & e Africa, Mid East, Asia, East Indies [Annular: Chad, Sudan, Ethiopia, Somalia, Myanmar, Thailand, Laos, Vietnam, se China]",12,9,729


**ANSWER:** The longest Solar Eclipse happened in December 14, 1955, with a total duration of 12 minutes and 9 seconds

In [ ]:
#When did the longest total solar eclipse occur?
split_duration.query('eclipse_type == "Total"').sort_values(by='total_duration', ascending=False).head(1)

,date,eclipse_type,magnitude,duration,region,minutes,seconds,total_duration
125,06-20-1955,Total,1.078,07m08s,"e Africa, e Asia, East Indies, n Australia [Total: Sri Lanka, Myanmar, Thailand, Laos, Cambodia, Vietnam, Philippines]",7,8,428


**ANSWER:** The longest Total Solar Eclipse happened in June 20, 1955, that lasts for 7 minutes and 8 seconds.

**QUESTION 2. What is the average duration of total solar eclipses?**

In [34]:
split_duration.query('eclipse_type == "Total"')['total_duration'].mean()

np.float64(211.96296296296296)

**ANSWER:** the average duration of total duration of total eclipses is approximately 211.96 seconds, or  3 minutes and 31 seconds.

**QUESTION 3. Show the next 10 solar eclipses?
	- Hint: convert date to datetime.**

In [ ]:
#converting date from str to datetime
df['date'] = pd.to_datetime(df['date'])

#current date: 08/04/2026
df.query('date >= "2026-08-04"').head(10)

,date,eclipse_type,magnitude,duration,region
277,2026-08-12,Total,1.039,02m18s,"n N. America, w Africa, Europe [Total: Arctic, Greenland, Iceland, Spain]"
278,2027-02-06,Annular,0.928,07m51s,"S. America, Antarctica, w & s Africa [Annular: Chile, Argentina, Atlantic]"
279,2027-08-02,Total,1.079,06m23s,"Africa, Europe, Mid East, w & s Asia [Total:Morocco, Spain, Algeria, Libya, Egypt, Saudi Arabia, Yemen, Somalia]"
280,2028-01-26,Annular,0.921,10m27s,"e N. America, C. & S. America, w Europe, nw Africa [Annular: Ecuador, Peru, Brazil, Suriname, Spain, Portugal]"
281,2028-07-22,Total,1.056,05m10s,"SE Asia, E. Indies, Australia, N.Z. [Total: Australia, N. Z.]"
282,2029-01-14,Partial,0.871,NaN,"N. America, C. America"
283,2029-06-12,Partial,0.458,NaN,"Arctic, Scandanavia, Alaska, n Asia, n Canada"
284,2029-07-11,Partial,0.230,NaN,"s Chile, s Argentina"
285,2029-12-05,Partial,0.891,NaN,"s Argentina, s Chile, Antarctica"
286,2030-06-01,Annular,0.944,05m21s,"Europe, n Africa, Mid East, Asia, Arctic, Alaska [Annular: Algeria, Tunisia, Greece, Turkey, Russia, n. China, Japan]"
